File: dev/js/flights-density.js
import * as vg from '../setup.js';

export default function(el) {
  const {
    Param, slider, vconcat, hconcat, plot, from,
    contour, raster, colorScale, colorScheme,
    xAxis, yAxis, xLabelAnchor, yLabelAnchor, xZero,
    marginLeft, marginRight, width, height,
  } = vg;

  const bandwidth = Param.value(7);
  const thresholds = Param.value(10);
  const table = 'flights';
  const x = 'time';
  const y = 'delay';

  el.appendChild(
    vconcat(
      hconcat(
        slider({ label: 'Bandwidth (σ)', as: bandwidth, min: 1, max: 100 }),
        slider({ label: 'Thresholds', as: thresholds, min: 2, max: 20 })
      ),
      plot(
        colorScale('symlog'), colorScheme('ylgnbu'),
        raster(from(table), { x, y, fill: 'density', bandwidth }),
        contour(from(table), { x, y, stroke: 'white', strokeOpacity: 0.5, bandwidth, thresholds }),
        xAxis('top'), xLabelAnchor('center'), xZero(true),
        yAxis('right'), yLabelAnchor('center'),
        width(700), height(500), marginLeft(5), marginRight(40)
      )
    )
  );
}

File: docs/public/specs/python/flights-density.py
import json
import vgplot as vg

meta = vg.meta(title="Flights Density", description="Density `heatmap` and `contour` lines for 200,000+ flights by departure hour and arrival delay. The sliders adjust the smoothing (bandwidth) and number of contour thresholds.\n")
data = vg.data(
    flights=vg.parquet("data/flights-200k.parquet")
)

view = vg.vconcat(
    vg.hconcat(
            vg.slider(label="Bandwidth (σ)", as_="$bandwidth", min=1, max=100),
            vg.slider(label="Thresholds", as_="$thresholds", min=2, max=20)
        ),
    vg.plot(
            vg.heatmap(data=vg.from_("flights"), x="time", y="delay", fill="density", bandwidth="$bandwidth"),
            vg.contour(data=vg.from_("flights"), x="time", y="delay", stroke="white", stroke_opacity=0.5, bandwidth="$bandwidth", thresholds="$thresholds"),
            vg.color_scale("symlog"),
            vg.color_scheme("ylgnbu"),
            vg.x_axis("top"),
            vg.x_label_anchor("center"),
            vg.x_zero(True),
            vg.y_axis("right"),
            vg.y_label_anchor("center"),
            vg.margin_top(30),
            vg.margin_left(5),
            vg.margin_right(40),
            vg.width(700),
            vg.height(500)
        )
)

params = {
    "bandwidth": 7,
    "thresholds": 10
}

spec = vg.spec(meta=meta, data=data, params=params, view=view)

if __name__ == "__main__":
    print(json.dumps(spec.to_dict(), sort_keys=True))